crypto/s2n_certificate.c

```diff
@@ -129,6 +135,34 @@ int s2n_cert_chain_and_key_set_private_key(struct s2n_cert_chain_and_key *cert_a
     return 0;
 }
 
+int s2n_cert_chain_and_key_set_private_key_using_len(struct s2n_cert_chain_and_key *cert_and_key, char *private_key_pem, uint32_t private_key_len)
+{
+    DEFER_CLEANUP(struct s2n_stuffer key_in_stuffer = {0}, s2n_stuffer_free);
+    DEFER_CLEANUP(struct s2n_stuffer key_out_stuffer = {0}, s2n_stuffer_free);
+
+    /* Put the private key pem in a stuffer */
+    POSIX_GUARD(s2n_stuffer_init_ro_from_string(&key_in_stuffer, (uint8_t *)private_key_pem, private_key_len));
+    POSIX_GUARD(s2n_stuffer_growable_alloc(&key_out_stuffer, private_key_len));
+
+    POSIX_GUARD(s2n_cert_chain_and_key_set_private_key_from_stuffer(cert_and_key, &key_in_stuffer, &key_out_stuffer));
+
+    return 0;
+}
+
+int s2n_cert_chain_and_key_set_private_key(struct s2n_cert_chain_and_key *cert_and_key, const char *private_key_pem)
+{
+    DEFER_CLEANUP(struct s2n_stuffer key_in_stuffer = {0}, s2n_stuffer_free);
+    DEFER_CLEANUP(struct s2n_stuffer key_out_stuffer = {0}, s2n_stuffer_free);
+
+    /* Put the private key pem in a stuffer */
+    POSIX_GUARD(s2n_stuffer_alloc_ro_from_string(&key_in_stuffer, private_key_pem));
+    POSIX_GUARD(s2n_stuffer_growable_alloc(&key_out_stuffer, strlen(private_key_pem)));
+
+    POSIX_GUARD(s2n_cert_chain_and_key_set_private_key_from_stuffer(cert_and_key, &key_in_stuffer, &key_out_stuffer));
+
+    return 0;
```

Comment:
to avoid duplicating code, can you simply:

```
return s2n_cert_chain_and_key_set_private_key_using_len(cert_and_key, private_key_pem, strlen(private_key_pem));
```

and do you need the `POSIX_ENSURE_REF` here?

Reply:
Yeah did that initially. But, s2n_cert_chain_and_key_load_pem takes const variables and s2n_blob doesn't use const. That's why removed const for the new api.

`extern int s2n_cert_chain_and_key_load_pem_using_len(struct s2n_cert_chain_and_key *chain_and_key, char *chain_pem, uint32_t chain_len, char *private_key_pem, uint32_t private_key_len); `

>and do you need the POSIX_ENSURE_REF here?
No. In s2n_stuffer_init_ro_from_string we do that check for the string (chain and pvt key).

Reply:
It would segfault. Why? Did I miss a place where this could happen?

Reply:
You do strlen(private_key_pem) in this method without verifying that private_key_pem!=NULL. That's why Matt is suggesting POSIX_ENSURE_REF.

Reply:
But, we do the POSIX_ENSURE_REF check in `s2n_stuffer_alloc_ro_from_string`. If it is null, s2n_stuffer_alloc_ro_from_string would return S2N_FAILURE. And with POSIX_GUARD we would return -1.

Reply:
This method shouldn't rely on null checking in another function, which could theoretically change without notice. If you dereference a pointer, you should verify that pointer isn't NULL.

Reply:
Added the check.